appsrc/acrnema/andump.man
.TH ANDUMP 1 "6 March 2007" "DICOM PS3" "DICOM PS3 - Describe ACR-NEMA file content"
.SH NAME
andump \- ACR/NEMA DICOM PS3 ... Describe ACR-NEMA file content
.SH SYNOPSIS
.HP 10
.B andump
[
.B \-ancreate
]
[
.B \-extractCSA
]
[
.B \-showoffset|-showoffset-hex
|-showoffset-octal|-showoffset-oct
|-showoffset-decimal|-showoffset-dec
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B andump
reads the named acr-nema input file and describes the information contained, without attempting to interpret the structure of the message (cf. dcdump).
.LP
The group and element number, dictionary and explicit value representation, description of tag, value length and value of the element are displayed, 
optionally with an offset byte count from the start of the file.
.SH OPTIONS
The attribute values, description and verbose output go to standard error.
.LP
Binary attributes are written in hexadecimal with a preceding
"0x". Numeric string attributes are written in decimal. Attribute values
are displayed in hexadecimal or string format as determined by the value representation.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-ancreate
.RS
Display item, end item and end sequence delimiter tags in a form suitable to feed into ancreate.
.RE
.TP
.B \-extractCSA
.RS
Extract the DICOM-like attributes buried within the Siemens private CSA header OB value.
.RE
.TP
.B \-showoffset|-showoffset-hex
.RS
Display byte offset from file start in hexadecimal.
.RE
.TP
.B \-showoffset-octal|-showoffset-oct
.RS
Display byte offset from file start in octal.
.RE
.TP
.B \-showoffset-decimal|-showoffset-dec
.RS
Display byte offset from file start in decimal.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% andump NM.dc3
.RE
(0x0008,0x0000) UL Group Length  VR=<UL> VL=<0x0004> [0x08] 
.RE
(0x0008,0x0008) CS Image Type    VR=<CS> VL=<0x0024> <...> 
.RE
...
.RE
(0x0054,0x0013) SQ ... Sequence  VR=<SQ> VL=<0xffffffff> []
.RE
(0xfffe,0xe000) NONE Item        VR=<> VL=<0xffffffff> []
.RE
(0x0054,0x0014) DS Energy Win LL VR=<DS> VL=<0x0006> <1.0 > 
.RE
(0x0054,0x0015) DS Energy Win UL VR=<DS> VL=<0x0006> <9.0 > 
.RE
(0xfffe,0xe00d) NONE Item Del .. VR=<> VL=<0x0000> []
.RE
(0xfffe,0xe0dd) NONE Seq Del ..  VR=<> VL=<0x0000> []
.RE
...
.RE
\ 
.RE
\ 
.RE
% andump -ancreate NM.dc3
.RE
(0x0008,0x0000) UL Group Length  VR=<UL> VL=<0x0004> [0x08] 
.RE
(0x0008,0x0008) CS Image Type    VR=<CS> VL=<0x0024> <...> 
.RE
...
.RE
(0x0054,0x0013) SQ ... Sequence  VR=<SQ> VL=<0xffffffff> []
.RE
%item
.RE
(0x0054,0x0014) DS Energy Win LL VR=<DS> VL=<0x0006> <1.0 > 
.RE
(0x0054,0x0015) DS Energy Win UL VR=<DS> VL=<0x0006> <9.0 > 
.RE
%enditem
.RE
%endseq
.RE
...
.RE
\ 
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR ancreate(1) ,
.BR dcdump(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/acrnema/andump.cc
#include <iostream.h>
#include <iomanip.h>
#include <string.h>
#include <ctype.h>

#ifdef POWINTEGEREXPONENTTYPE
#define	powi(m,e)	pow(m,(POWINTEGEREXPONENTTYPE)(e))
#else
#define	powi(m,e)	pow(m,(long)(e))
#endif

#ifdef __SC__
#define infinity()    HUGE_VAL
#define quiet_nan(x)  0
#endif /* __SC__ */

#ifdef __MWERKS__
#define infinity()    __inf()
#define quiet_nan(x)  0
#endif /* __MWERKS__ */

#ifdef USEDUMBINFINITYANDNAN
#if USEDUMBINFINITYANDNAN
#define infinity()    HUGE_VAL
#define quiet_nan(x)  0
#endif
#endif

#include "attrtag.h"
#include "dcstream.h"
#include "dcopt.h"
#include "elmconst.h"
#include "elmdict.h"
#include "mesgtext.h"

#ifndef MAXVLTODUMP
#define MAXVLTODUMP 32768
#endif

static bool
isOtherByteOrWordVR(const char *vr)
{
	return vr && vr[0]=='O' && (vr[1]=='B' || vr[1]=='W' || vr[1]=='X');
}

static bool
isOtherByteOrWordOrFloatVR(const char *vr)
{
	return vr && vr[0]=='O' && (vr[1]=='B' || vr[1]=='F'  || vr[1]=='W' || vr[1]=='X');
}

static bool
isSequenceVR(const char *vr)
{
	return vr && vr[0]=='S' &&  vr[1]=='Q';
}

static bool
isUnknownVR(const char *vr)
{
	return vr && vr[0]=='U' &&  vr[1]=='N';
}

static bool
isUnlimitedTextVR(const char *vr)
{
	return vr && vr[0]=='U' &&  vr[1]=='T';
}

static TextOutputStream& 
writebase(TextOutputStream& stream,ElementDictionary& dict,const Tag tag,const char *vr,Uint32 vl)
{
	tag.write(stream,&dict);
	stream << "\t VR=<" << (vr ? vr : "") << ">  " << flush;
	stream << " VL=<";
	writeZeroPaddedHexNumber(stream,vl,4);
	stream << ">  " << flush;
	return stream;
}

static bool
isStringVR(const char *vr)
{
	return vr &&
		(  strcmp(vr,"AE") == 0
		|| strcmp(vr,"AS") == 0
		|| strcmp(vr,"CS") == 0
		|| strcmp(vr,"DA") == 0
		|| strcmp(vr,"DT") == 0
		|| strcmp(vr,"DS") == 0
		|| strcmp(vr,"IS") == 0
		|| strcmp(vr,"LO") == 0
		|| strcmp(vr,"LT") == 0
		|| strcmp(vr,"PN") == 0
		|| strcmp(vr,"SH") == 0
		|| strcmp(vr,"ST") == 0
		|| strcmp(vr,"TM") == 0
		|| strcmp(vr,"UI") == 0
		|| strcmp(vr,"UT") == 0);
}

static bool
isNumericVR(const char *vr)
{
	return vr &&
		(  strcmp(vr,"OB") == 0
		|| strcmp(vr,"OW") == 0
		|| strcmp(vr,"OX") == 0
		|| strcmp(vr,"SL") == 0
		|| strcmp(vr,"SS") == 0
		|| strcmp(vr,"UL") == 0
		|| strcmp(vr,"US") == 0
		|| strcmp(vr,"XL") == 0
		|| strcmp(vr,"XS") == 0);
}

static bool
isFloatVR(const char *vr)
{
	return vr &&
		(  strcmp(vr,"FL") == 0
		|| strcmp(vr,"FD") == 0);
}

static bool
isTagVR(const char *vr)
{
	return vr && strcmp(vr,"AT") == 0;
}

static size_t
sizeofNumericVR(const char *vr)
{
	if (     strcmp(vr,"OB") == 0
	      || strcmp(vr,"OX") == 0)
		return 1;
	else if (strcmp(vr,"OW") == 0
	      || strcmp(vr,"US") == 0
	      || strcmp(vr,"SS") == 0
	      || strcmp(vr,"XS") == 0)
		return 2;
	else if (strcmp(vr,"SL") == 0
	      || strcmp(vr,"UL") == 0
	      || strcmp(vr,"XL") == 0)
		return 4;
	else {
		Assert(0);
	}
	return 0;	// never gets here but prevents warning
}

static size_t
sizeofFloatVR(const char *vr)
{
	if (     strcmp(vr,"FL") == 0)
		return 4;
	else if (strcmp(vr,"FD") == 0)
		return 8;
	else {
		Assert(0);
	}
	return 0;	// never gets here but prevents warning
}

static TextOutputStream&
doStringValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl,const char * &rstr)
{
	static char buffer[MAXVLTODUMP];
	Assert(vl < MAXVLTODUMP);
	if (vl) in.read(buffer,(int)vl);
	buffer[vl]=0;
	if (!in.fail()) {
		log << "<";
		char *ptr=buffer;
		while (vl--) {
			if (*ptr)
				log << *ptr;
			else if (vl) {	// Allow single trailing null eg. for UI string
				log << endl << WMsgDC(UnexpectedNullInString) << endl;
				break;
			}
			++ptr;
		}
		log << "> ";
		// return value has no leading or trailing space
		ptr=buffer+strlen(buffer);
		while (ptr >= buffer && (*ptr == 0 || isspace(*ptr))) *ptr--=0;
		rstr=buffer;
		while (*rstr && isspace(*rstr)) ++rstr;
	}
	else {
		cerr << EMsgDC(ReadFailed) << endl;
		rstr=0;
	}
	return log;
}

static TextOutputStream&
doTagValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl)
{
	if (vl%4) {
		cerr << EMsgDC(BadValueLength) << endl;
		in.seekg(vl,ios::cur);
		if (in.fail()) {
			cerr << EMsgDC(SeekFailed) << endl;
		}
	}
	else {
		log << "{";
		while (vl) {
			Uint16 group;
			Uint16 element;
			in >> group;
			in >> element;
			log << "(";
			writeZeroPaddedHexNumber(log,group,4);
			log << ",";
			writeZeroPaddedHexNumber(log,element,4);
			log << ")";
			vl-=4;
		}
		log << "} ";
		if (in.fail()) {
			cerr << EMsgDC(ReadFailed) << endl;
		}
	}
	return log;
}

static TextOutputStream&
doBinaryValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl,size_t size,Uint32& rval)
{
	rval=0;
	if (vl < size) size=vl;	// handles special case of 1 US where UL wanted
	if (vl%size) {
		cerr << EMsgDC(BadValueLength) << endl;
		in.seekg(vl,ios::cur);
		if (in.fail()) {
			cerr << EMsgDC(SeekFailed) << endl;
		}
	}
	else {
		log << "[";
		while (vl) {
			Uint32 value;
			switch(size) {
				case 1: { Uint8  v; in >> v; value=v; } break;
				case 2: { Uint16 v; in >> v; value=v; } break;
				case 4: { Uint32 v; in >> v; value=v; } break;
				default: Assert(0); break;
			}
			rval=value;
			writeZeroPaddedHexNumber(log,value,size*2);
			vl-=size;
			if (vl) log << ",";
		}
		log << "] ";
		if (in.fail()) {
			cerr << EMsgDC(ReadFailed) << endl;
		}
	}
	return log;
}

static TextOutputStream&
doFloatValues(DicomInputStream& in,TextOutputStream& log,Uint32 vl,size_t size)
{
	if (vl%size) {
		Uint32 ival;
		doBinaryValues(in,log,vl,1,ival);
		cerr << EMsgDC(BadValueLength) << endl;
		//in.seekg(vl,ios::cur);
		//if (in.fail()) {
		//	cerr << EMsgDC(SeekFailed) << endl;
		//}
	}
	else {
		log << "{";
		while (vl) {
			if (size == 4) {
				Uint32 binary; in >> binary;

				double value;
				Int16 sign	=(Int16)((binary&0x80000000)>>31);
				Int16 exponent	=(Int16)((binary&0x7f800000)>>23);
				Uint32 mantissa	= binary&0x007fffff;

				if (exponent == 0) {
					if (mantissa == 0)
						value=0;
					else {
						value=((double)mantissa/
							(1<<23))*powi(2.0,-126);
						value = (sign == 0) ? value : -value;
					}
				}
				else if (exponent == 255) {
					if (mantissa)
						value=quiet_nan(0);
					else
						value=infinity();
				}
				else {
					value=(1.0+(double)mantissa/
						(1<<23))*powi(2.0,exponent-127);
					value = (sign == 0) ? value : -value;
				}
				log << dec << value;
			}
			else if (size == 8) {
				Uint32 binary1; in >> binary1;
				Uint32 binary2; in >> binary2;

				double value;
				Uint32 high = in.isBigEndian() ? binary1 : binary2;
				Uint32 low  = in.isBigEndian() ? binary2 : binary1;

				Int16 sign	=(Int16)((high&0x80000000)>>31);
				Int16 exponent	=(Int16)((high&0x7ff00000)>>20);
				Uint32 mantissahigh = high&0x000fffff;
				Uint32 mantissalow  = low &0xffffffff;
				double mantissavalue = (double)mantissahigh * powi(2.0,32)
						     + (double)mantissalow;

				if (exponent == 0) {
					if (mantissahigh == 0 && mantissalow == 0)
						value=0;
					else {
						value=(mantissavalue/powi(2.0,52))
							*powi(2.0,-1022);
						value = (sign == 0) ? value : -value;
					}
				}
				else if (exponent == 255) {
					if (mantissahigh || mantissalow)
						value=quiet_nan(0);
					else
						value=infinity();
				}
				else {
					value=(1.0+mantissavalue/powi(2.0,52))
						*powi(2.0,exponent-1023);
					value = (sign == 0) ? value : -value;
				}
				log << dec << value;
			}
			else {
				Assert(0);
			}

			vl-=size;
			if (vl) log << ",";
		}
		log << "} ";
		if (in.fail()) {
			cerr << EMsgDC(ReadFailed) << endl;
		}
	}
	return log;
}

static bool
readAll(DicomInputStream& stream,TextOutputStream& log,ElementDictionary& dict,bool showoffset,ios::fmtflags showoffsetbase,bool ancreate,bool extractCSA)
{
	Uint32 n=stream.haveMetaHeader() ? 132 : 0;
	Uint32 length=0xffffffff;
	bool lastwasencapOXorItem=false;
	bool donewithmetaheader=false;
	Tag *siemensCSAHeaderInfoTag = NULL;
	while (stream.peek() != EOF && (length == 0xffffffff || n < length)) {
		if (showoffset) {
			log << "@";
			writeZeroPaddedNumber(log,n,showoffsetbase,8);
#ifdef CHECKFORSGINATIVECCSEEKBUG
			log << "(";
			writeZeroPaddedHexNumber(log,stream.rdbuf()->PubSeekOff(0,ios::cur),8);
			log << ")";
#endif
			log << ": ";
		}
		Tag tag;
		stream >> tag;
		if (stream.fail()) {
			cerr << EMsgDC(TagReadFailed) << endl;
			return false;
		}
		n+=4;

		if (!donewithmetaheader && !tag.isMetaheaderGroup()) {
			if (stream.getTransferSyntaxToReadDataSet()) {
				stream.readingDataSet();
				donewithmetaheader=true;
				// backup and read again in case endian changed ...
#ifdef CHECKFORSGINATIVECCSEEKBUG
log << endl << "Before endian check backup @";
writeZeroPaddedHexNumber(log,n,8);
log << "(";
writeZeroPaddedHexNumber(log,stream.rdbuf()->PubSeekOff(0,ios::cur),8);
log << ")"
    << ": ";
#endif
				stream.seekg(-4l,ios::cur);
				if (stream.fail()) {
					cerr << EMsgDC(TagReadFailed) << " - " << EMsgDC(SeekFailed) << endl;
					return false;
				}
#ifdef CHECKFORSGINATIVECCSEEKBUG
log << endl << "Before endian check backup @";
writeZeroPaddedHexNumber(log,n,8);
log << "(";
writeZeroPaddedHexNumber(log,stream.rdbuf()->PubSeekOff(0,ios::cur),8);
log << ")"
    << ": " << endl;
#endif
				stream >> tag;
				if (stream.fail()) {
					cerr << EMsgDC(TagReadFailed) << endl;
					return false;
				}
			}
			else {
				cerr << EMsgDC(NoTransferSyntaxInMetaHeader) << endl;
				return false;
			}
		}

		char vre[3];
		const char *vr;
		Uint32 vl;

		if (tag == TagFromName(ItemDelimitationItem)
		 || tag == TagFromName(SequenceDelimitationItem)
		 || tag == TagFromName(Item)
		 || tag == Tag(0xfeff,0x00e0)				// GE bug (endianness of Item tag in encapsulated pixel data)
		 || tag == Tag(0xfeff,0xdde0)				// GE bug (endianness of Sequence Delimitation Item tag in encapsulated pixel data)
		) {
			vr=0;
			stream >> vl;
			if (tag == Tag(0xfeff,0x00e0)) {
				vl = ((vl>>24)&0xff) + (((vl>>16)&0xff)<<8) + (((vl>>8)&0xff)<<16) +((vl&0xff)<<24);	// GE bug - swap VL endianness as well
			}
			n+=4;
		}
		else  {
			if (stream.getTransferSyntaxInUse()->isExplicitVR()) {
				stream.read(vre,2);
				if (stream.fail()) {
					cerr << EMsgDC(VRReadFailed) << endl;
					return false;
				}
				if (vre[0] < 'A' || vre[1] < 'A') {		// DicomWorks bug ... occasional implicit VR attribute even though explicit transfer syntax
					//cerr << EMsgDC(TagReadFailed) << " - " << MMsgDC(ImplicitVREncodingWhenExplicitRequired) << endl;
					vr="UN";
					stream.seekg(-2l,ios::cur);
					if (stream.fail()) {
						cerr << EMsgDC(TagReadFailed) << " - " << EMsgDC(SeekFailed) << endl;
						return false;
					}
					stream >> vl;				// read VL as if implicit VR
					n+=4;
				}
				else {
					n+=2;
					vre[2]=0;
					vr=vre;
					if (isOtherByteOrWordOrFloatVR(vr) || isSequenceVR(vr)) {
						(void)stream.read16();	// "Reserved"
						stream >> vl;
						n+=6;
					}
					else if (isUnlimitedTextVR(vr) || isUnknownVR(vr)) {
						// should be long form, but check for bad implementations that use short form ...
						vl=stream.read16();
						n+=2;
						if (vl == 0) {		// if reserved bytes not zero, assume used for short form of length (risky) :(
							stream >> vl;
							n+=4;
						}
						else {
							cerr << EMsgDC(IncorrectShortVLForUTOrUN) << endl;
						}
					}
					else {
						vl=stream.read16();
						n+=2;
					}
				}
			}
			else {
				vr=(tag.isPrivateGroup() && tag.isPrivateOwner())
					? "LO"	// PS3.5-7.8.1
					: dict.getValueRepresentation(tag);
				stream >> vl;
				n+=4;
			}
		}
		if (stream.fail()) {
			cerr << EMsgDC(VLReadFailed) << endl;
			return false;
		}

		if (ancreate) {
			if (tag == TagFromName(ItemDelimitationItem)) {
				log << "%enditem";
			}
			else if (tag == TagFromName(SequenceDelimitationItem)) {
				log << "%endseq";
			}
			else if (tag == TagFromName(Item)) {
				log << "%item";
			}
			else {
				writebase(log,dict,tag,vr,vl);
			}
		}
		else {
			writebase(log,dict,tag,vr,vl);
		}
		
		if (vl == 0xffffffff		// SQ or OX both will contain items
		 || vl == 0			// includes ItemDelimitationItem,SequenceDelimitationItem
		 || isSequenceVR(vr)		// SQ with defined length
		 || (tag == TagFromName(Item) && !lastwasencapOXorItem)	// Item with element contents
		) {
			if (!ancreate
			 || (tag != TagFromName(ItemDelimitationItem)
			  && tag != TagFromName(SequenceDelimitationItem)
			  && tag != TagFromName(Item))) {
				log << "\t[]";
			}
		}
		else {
			if (extractCSA && siemensCSAHeaderInfoTag && *siemensCSAHeaderInfoTag == tag) {
				// do nothing - treat it as containing DICOM tags and loop around - works for EVRLE only
				// should also check to stop at (0xffff,0xffff)
			}
			else if (vl < MAXVLTODUMP) {
				if (isStringVR(vr)) {
					const char *sval;
					doStringValues(stream,log,vl,sval);
					if (!stream.fail() && vl && sval && *sval) {
						if (tag == TagFromName(TransferSyntaxUID)) {
							if (!donewithmetaheader) {
								stream.setTransferSyntaxToReadDataSet(new TransferSyntax(sval));
							}
						}
						else if (tag.isPrivateGroup() && tag.isPrivateOwner()) {
							dict.addOwner(tag,sval);
							if (tag.getGroup() == 0x0029 && strcmp(sval,"SIEMENS CSA HEADER") == 0) {
								Uint16 siemensCSAHeaderInfoElement=((tag.getElement() << 8) & 0xff00)+0x0010;
								siemensCSAHeaderInfoTag = new Tag(0x0029,siemensCSAHeaderInfoElement);
							}
						}
					}
				}
				else if (isTagVR(vr))
					doTagValues(stream,log,vl);
				else if (isNumericVR(vr)) {
					Uint32 ival;
					size_t size=sizeofNumericVR(vr);
					doBinaryValues(stream,log,vl,size,ival);
					if (!stream.fail() && vl == size) {
						if (tag == TagFromName(LengthToEnd)
						 && length == 0xffffffff) {
							// Retired, but if you've got it use it
							//length=n+ival+vl;
						}
					}
				}
				else if (isFloatVR(vr)) {
					size_t size=sizeofFloatVR(vr);
					doFloatValues(stream,log,vl,size);
				}
				else {
					//cerr << endl << WMsgDC(UnrecognizedVR) << endl;
					Uint32 ival;
					doBinaryValues(stream,log,vl,1,ival);
				}
			}
			else {
				Assert(!ancreate);
				log << "\t[]\t# skipping ...";		
				stream.seekg(vl,ios::cur);
				if (stream.fail()) {
					cerr << EMsgDC(SeekFailed) << endl;
					return false;
				}
			}
			n+=vl;
		}
		lastwasencapOXorItem=
			(isOtherByteOrWordVR(vr)
			 || tag == TagFromName(Item)
			 || tag == TagFromName(ItemDelimitationItem)
			) && stream.getTransferSyntaxInUse()->isEncapsulated();

		log << endl;
	}
	return true;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions    options(argc,argv);
	DicomInputOptions input_options(options);

	bool showfilename=options.get("filename");
	bool ancreate=options.get("ancreate");
	bool extractCSA=options.get("extractCSA");
	bool showoffset=false;
	ios::fmtflags showoffsetbase;
	if (options["showoffset"]||options["showoffset-hex"])
		{ showoffset=true; showoffsetbase=ios::hex; }
	else if (options["showoffset-octal"]||options["showoffset-oct"])
		{ showoffset=true; showoffsetbase=ios::oct; }
	else if (options["showoffset-decimal"]||options["showoffset-dec"])
		{ showoffset=true; showoffsetbase=ios::dec; }

	input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,input_options.filename,cin);

	cerr << options.errors();
	cerr << input_options.errors();
	cerr << input_opener.errors();

	if (!input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< input_options.usage()
			<< " [-ancreate]"
			<< " [-extractCSA]"
			<< " [-showoffset|-showoffset-hex"
			<< "|-showoffset-octal|-showoffset-oct"
			<< "|-showoffset-decimal|-showoffset-dec]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		input_options.transfersyntaxuid,
		input_options.usemetaheader);
	TextOutputStream log(cerr);

	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		log << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	ElementDictionary dictionary;

	readAll(din,log,dictionary,showoffset,showoffsetbase,ancreate,extractCSA);
}
